### PR TITLE
Ide 4584

### PR DESCRIPTION
**Pull request by jtydhr88**

### CODE REVIEW

**Comment by gamerson** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: tools/plugins/com.liferay.ide.server.tomcat.core/src/com/liferay/ide/server/tomcat/core/PortalContext.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ */
+
+package com.liferay.ide.server.tomcat.core;
+
+/**
+ * @author Seiphon Wang
+ */
+public class PortalContext {
+
+	public PortalContext(String name) {
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.server.tomcat.core/src/com/liferay/ide/server/tomcat/core/PortalContext.java]: Can you describe the testing environment that shows how this class works?  can you add a unit test for this functionality?

File: tools/plugins/com.liferay.ide.server.core/src/com/liferay/ide/server/core/portal/PortalServerBehavior.java [hunk 2]
```diff
@@ -210,6 +210,18 @@ public void launchServer(ILaunch launch, String mode, IProgressMonitor monitor)
 				url += ":" + port;
 			}
 
+			IPath appServerPortalPath = _getPortalRuntime().getAppServerPortalDir();
+
+			if (appServerPortalPath != null) {
+				String appServerPortalDir = appServerPortalPath.toString();
+
+				String folderName = appServerPortalDir.substring(appServerPortalDir.lastIndexOf("/"));
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.server.core/src/com/liferay/ide/server/core/portal/PortalServerBehavior.java]: Is the context name always equals to the directory name?  in tomcat how do you specify the custom context name?  Through a directoyr name or through a context.xml file?

File: tools/plugins/com.liferay.ide.server.tomcat.core/src/com/liferay/ide/server/tomcat/core/PortalTomcatBundle.java [hunk 3]
```diff
@@ -77,13 +79,35 @@ public IPath getAppServerLibGlobalDir() {
 
 	@Override
 	public IPath getAppServerPortalDir() {
-		IPath retval = null;
+		IPath appBasePath = bundlePath.append("webapps");
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.server.tomcat.core/src/com/liferay/ide/server/tomcat/core/PortalTomcatBundle.java]: This method now performs a calculation for each time it is called.  Instead it should inintialize a appServerPortalDIr the first time it is accessed but then should be cached there-after.  If the bundlePath changes, then it should re-calculate.

File: tools/plugins/com.liferay.ide.server.core/src/com/liferay/ide/server/core/portal/PortalServerBehavior.java (hunk 2 above)
Comment (Seiphon) [reply tools/plugins/com.liferay.ide.server.core/src/com/liferay/ide/server/core/portal/PortalServerBehavior.java]: Hi Greg, in tomcat we can specify custom context name by both of them.

Actually, when deploy applications, Tomcat will always process XML descriptors first, then packaged WARs, and then exploded WARs. And how to get context name from them, that is the work that "PortalContext" class will do. It called "ContextName" in tomcat, it creates an instance from a context name, display name, base name, directory name, War name or context.xml name. Then we can get context name or context path by get method.

**Pull request closed without merging**